sopel/bot.py

```diff
@@ -592,6 +592,58 @@ def register_urls(self, urls: Iterable) -> None:
                 except plugins.exceptions.PluginError as err:
                     LOGGER.error("Cannot register URL callback: %s", err)
 
+    def rate_limit_info(self, rule: plugin_rules.Rule, trigger: Trigger) -> Tuple[bool, Optional[str]]:
+        if trigger.admin or rule.is_unblockable():
+            return False, None
+
+        is_channel = trigger.sender and not trigger.sender.is_nick()
+        channel = trigger.sender if is_channel else 'private-message'
+
+        at_time = trigger.time
+
+        user_metrics = rule.get_user_metrics(trigger.nick)
+        channel_metrics = rule.get_channel_metrics(channel)
+        global_metrics = rule.get_global_metrics()
+
+        if user_metrics.is_limited(at_time - rule.user_rate_limit):
+            template = rule.user_rate_template
+            rate_limit_type = "user"
+            rate_limit = rule.user_rate_limit
+            metrics = user_metrics
+        elif is_channel and channel_metrics.is_limited(at_time - rule.channel_rate_limit):
+            template = rule.channel_rate_template
+            rate_limit_type = "channel"
+            rate_limit = rule.channel_rate_limit
+            metrics = channel_metrics
+        elif global_metrics.is_limited(at_time - rule.global_rate_limit):
+            template = rule.global_rate_template
+            rate_limit_type = "global"
+            rate_limit = rule.global_rate_limit
+            metrics = global_metrics
+        else:
+            return False, None
+
+        next_time = metrics.last_time + rate_limit
+        time_left = next_time - at_time
+
+        if template:
+            message = template.format(
+                nick=trigger.nick,
```

Comment:
`trigger.nick` being used twice would mean an intermediate variable to me.

Reply:
(Not for me, though. Attributes are meant to be accessed!)

Reply:
Usage twice seems like a low threshold to me, I usually define a variable in the case that either:

1) it helps the legibility (i.e. because the expr being replaced is long/complex)
2) the attribute is likely to change between retrievals

But I'm not sure I see either of those cases here. Willing to change it though, I'll let you two work that one out :sweat_smile:

Reply:
It's a bit weird to have `channel` but not `nick` tbh. 😁 

Other than that, yeah, it's fine.

Reply:
> It's a bit weird to have `channel` but not `nick` tbh. grin

If `Trigger` provided a `channel` attribute, I would have happily used it here and would happily do so in many other places too :stuck_out_tongue_winking_eye: 

In this case, I defined `channel` because the expression that checks `trigger.sender` _is_ a burden to repeat and would harm the legibility of surrounding code